runtime/near-vm-errors/src/lib.rs

```diff
@@ -186,6 +186,8 @@ pub enum HostError {
     ContractSizeExceeded { size: u64, limit: u64 },
     /// The host function was deprecated.
     Deprecated { method_name: String },
+    /// Invalid ECDSA signature.
+    InvalidECDSASignature,
```

Comment:
Should it also be feature-guarded?

Reply:
Yeah looks like it should be @artob

Reply:
Wait, should we have this at all? As per @abacabadabacaba  comment, `ecrecover` should return an error, rather than abort execution. My understanding is that, if we emit an `HostError`, we unrecoverably abort transaction?

I might be misunderstading something, but I think what we need here is something like

```diff
diff --git a/runtime/near-vm-logic/src/logic.rs b/runtime/near-vm-logic/src/logic.rs
index 9bb74980..aa5cc40e 100644
--- a/runtime/near-vm-logic/src/logic.rs
+++ b/runtime/near-vm-logic/src/logic.rs
@@ -1097,10 +1097,10 @@ impl<'a> VMLogic<'a> {
         r_ptr: u64,
         s_ptr: u64,
         register_id: u64,
-    ) -> Result<()> {
+    ) -> Result<bool> {
         self.gas_counter.pay_base(ecrecover_base)?;
         if v != 27 && v != 28 {
-            return Err(HostError::InvalidECDSASignature.into());
+            return Ok(false);
         }
         let hash = self.memory_get_vec(hash_ptr, 32)?;
         let v = (v & 0xFF) as u8;
@@ -1115,12 +1115,12 @@ impl<'a> VMLogic<'a> {
         let signature = secp256k1::Signature::parse(&signature);
 
         let recovery_id = match secp256k1::RecoveryId::parse_rpc(v) {
-            Err(_) => return Err(HostError::InvalidECDSASignature.into()),
+            Err(_) => return Ok(false),
             Ok(rid) => rid,
         };
 
         let public_key = match secp256k1::recover(&hash, &signature, &recovery_id) {
-            Err(_) => return Err(HostError::InvalidECDSASignature.into()),
+            Err(_) => return Ok(false),
             Ok(pk) => pk,
         };
 
@@ -1129,7 +1129,8 @@ impl<'a> VMLogic<'a> {
         let mut address = [0u8; 20];
         address.copy_from_slice(&result[12..]);
 
-        self.internal_write_register(register_id, address.to_vec())
+        self.internal_write_register(register_id, address.to_vec());
+        Ok(true)
     }
 
     /// Called by gas metering injected into Wasm. Counts both towards `burnt_gas` and `used_gas`.
```

?

Reply:
Sounds like @matklad  is right.

Btw, it is extremely important to categorize error messages correctly with the runtime. Almost any incorrect miscategorization will either lead to a network split/slashing or Mainnet shutdown.